File: Maximum_Profit_in_Job_Scheduling.java
class Solution {
    public int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
        int max_profit = 0;
        int n = startTime.length;
        int[][] jobs = new int[n][3];
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[0] - b[0]);

        for (int i = 0; i < n; i++) {
            jobs[i] = new int[]{startTime[i], endTime[i], profit[i]};
        }

        Arrays.sort(jobs, (a, b) -> a[0] - b[0]);

        for (int[] job : jobs) {
            while (!heap.isEmpty() && job[0] >= heap.peek()[0]) {
                max_profit = Math.max(max_profit, heap.poll()[1]);
            }
            heap.offer(new int[]{job[1], job[2] + max_profit});
        }

        return Math.max(max_profit, heap.stream().mapToInt(arr -> arr[1]).max().orElse(0));        
    }
}

File: Maximum_Profit_in_Job_Scheduling.cpp
class Solution {
public:
    int jobScheduling(vector<int>& startTime, vector<int>& endTime, vector<int>& profit) {
        int max_profit = 0;
        int n = startTime.size();
        vector<vector<int>> jobs(n, vector<int>(3));
        priority_queue<vector<int>, vector<vector<int>>, greater<>> heap;

        for (int i = 0; i < n; i++) {
            jobs[i] = {startTime[i], endTime[i], profit[i]};
        }

        sort(jobs.begin(), jobs.end(), [](const auto& a, const auto& b) {
            return a[0] < b[0];
        });

        for (const auto& job : jobs) {
            while (!heap.empty() && job[0] >= heap.top()[0]) {
                max_profit = max(max_profit, heap.top()[1]);
                heap.pop();
            }
            heap.push({job[1], job[2] + max_profit});
        }

        while (!heap.empty()) {
            max_profit = max(max_profit, heap.top()[1]);
            heap.pop();
        }

        return max_profit;        
    }
};
